Remove some unnecessary regex NonBacktracking code

stephentoub opened the pull request:
Move some debug-only code to #if DEBUG, remove other dead code, and clean up a few things that are unnecessary.  Also add a few tests to fill out some code coverage results, and address some style-related comments from a recently merged PR.

olsaarik commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/DoublyLinkedList.cs:
```diff
@@ -14,7 +14,7 @@ namespace System.Text.RegularExpressions.Symbolic
     /// where it maintains linear construction time in terms of the overall number of AST nodes in a given <see cref="RegexNode"/> input.
     /// </remarks>
-    internal sealed class DoublyLinkedList<T> where T : notnull
+    internal sealed class DoublyLinkedList<T> : IEnumerable<T> where T : notnull
```
A remark that the `IEnumerable` is implemented in reverse might help avoid future confusion here.

olsaarik commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexBuilder.cs:
```diff
@@ -118,41 +118,16 @@ internal sealed class SymbolicRegexBuilder<TSet> where TSet : IComparable<TSet>,
-        internal (bool IsInitial, bool IsDeadend, bool IsNullable, bool CanBeNullable) GetStateInfo(int stateId)
+        /// <summary>Context-independent information available for every state.</summary>
+        [Flags]
+        private enum ContextIndependentState : byte
```
Oh nice. This is much cleaner.

olsaarik reviewed (approved):
Looks good!

stephentoub committed:
Remove some unnecessary regex NonBacktracking code

Move some debug-only code to #if DEBUG, remove other dead code, and clean up a few things that are unnecessary.  Also add a few tests to fill out some code coverage results, and address some style-related comments from a recently merged PR.

stephentoub commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexBuilder.cs:
```diff
@@ -253,17 +271,33 @@ internal TSet GetMinterm(int mintermId)
-            int numMinterms = state.StartsWithLineAnchor ? _minterms.Length + 1 : _minterms.Length;
+            {
+                return default;
+            }
+
+            int numMinterms = _minterms.Length;
+            if (state.StartsWithLineAnchor)
+            {
+                numMinterms++;
+            }
+
             return _delta.AsSpan(state.Id << _mintermsLog, numMinterms);
         }
 
         /// <summary>Returns the span from <see cref="_nfaDelta"/> that may contain transitions for the given state</summary>
         internal Span<int[]?> GetNfaDeltasFor(DfaMatchingState<TSet> state)
         {
             if (_nfaDelta is null || _minterms is null || !_nfaStateArrayInverse.TryGetValue(state.Id, out int nfaState))
-                return Span<int[]?>.Empty;
-            int numMinterms = state.StartsWithLineAnchor ? _minterms.Length + 1 : _minterms.Length;
+            {
+                return default;
+            }
+
+            int numMinterms = _minterms.Length;
+            if (state.StartsWithLineAnchor)
+            {
+                numMinterms++;
+            }
+
```
Thanks.  I'll leave this one as-is for now, since I don't fully understand the code yet.

stephentoub committed:
Address PR feedback

stephentoub committed:
Fix non-deterministic threading-related failures in regex tests

If the wrong tests run concurrently, they can end up seeing a changed NFA safe limit from another test.
